File: app/src/main/java/me/devsaki/hentoid/json/JsonImageFile.java
package me.devsaki.hentoid.json;

import androidx.annotation.NonNull;

import com.annimon.stream.Optional;
import com.annimon.stream.Stream;

import java.util.List;

import me.devsaki.hentoid.database.domains.Chapter;
import me.devsaki.hentoid.database.domains.ImageFile;
import me.devsaki.hentoid.enums.StatusContent;

class JsonImageFile {

    private Integer order;
    private String url;
    private String name;
    private boolean isCover;
    private boolean favourite;
    private boolean isRead;
    private StatusContent status;
    private String mimeType;
    private long pHash;

    private int chapterOrder = -1;

    private JsonImageFile() {
    }

    static JsonImageFile fromEntity(ImageFile f) {
        JsonImageFile result = new JsonImageFile();
        result.order = f.getOrder();
        result.url = f.getUrl();
        result.name = f.getName();
        result.status = f.getStatus();
        result.isCover = f.isCover();
        result.favourite = f.isFavourite();
        result.isRead = f.isRead();
        result.mimeType = f.getMimeType();
        result.pHash = f.getImageHash();
        if (f.getChapter() != null && f.getChapter().getTarget() != null)
            result.chapterOrder = f.getChapter().getTarget().getOrder();
        return result;
    }

    ImageFile toEntity(int maxPages, @NonNull List<Chapter> chapters) {
        ImageFile result = new ImageFile(order, url, status, maxPages);
        result.setName(name);
        result.setIsCover(isCover);
        result.setFavourite(favourite);
        result.setRead(isRead);
        result.setMimeType(mimeType);
        result.setImageHash(pHash);

        if (!chapters.isEmpty() && chapterOrder > -1) {
            Optional<Chapter> chapter = Stream.of(chapters).filter(c -> c.getOrder().equals(chapterOrder)).findFirst();
            if (chapter.isPresent()) result.setChapter(chapter.get());
        }

        return result;
    }
}

File: app/src/main/java/me/devsaki/hentoid/json/JsonImageFile.kt
package me.devsaki.hentoid.json

import com.squareup.moshi.JsonClass
import me.devsaki.hentoid.database.domains.Chapter
import me.devsaki.hentoid.database.domains.ImageFile
import me.devsaki.hentoid.enums.StatusContent

@JsonClass(generateAdapter = true)
data class JsonImageFile(
    val order: Int,
    val url: String,
    val pageUrl: String?,
    val name: String,
    val isCover: Boolean?,
    val favourite: Boolean?,
    val isRead: Boolean?,
    val status: StatusContent,
    val pHash: Long?,
    val isTransformed: Boolean?,
    val chapterOrder: Int?
) {
    constructor(f: ImageFile) : this(
        f.order,
        f.url,
        f.pageUrl,
        f.name,
        f.isCover,
        f.favourite,
        f.read,
        f.status,
        f.imageHash,
        f.isTransformed,
        f.linkedChapter?.order ?: -1
    )

    fun toEntity(chapters: List<Chapter>): ImageFile {
        val result = if (url.isEmpty()) ImageFile.fromPageUrl(order, pageUrl ?: "", status, name)
        else ImageFile.fromImageUrl(order, url, status, name)
        result.name = name
        result.isCover = isCover == true
        result.favourite = favourite == true
        result.read = isRead == true
        result.imageHash = pHash ?: 0
        result.isTransformed = isTransformed == true

        if (chapters.isNotEmpty() && (chapterOrder ?: -1) > -1) {
            chapters.firstOrNull { it.order == chapterOrder }?.let {
                result.setChapter(it)
                it.imageFiles.add(result)
            }
        }

        return result
    }
}